src/dinit-iostream.cc

```diff
@@ -0,0 +1,740 @@
+{
+    return (num == 0) ? pow : type_max_numdigits(num / 10, pow + 1);
+}
+
+// We should have space for sign in signed variables.
+constexpr unsigned UINT_MAX_CHARS = type_max_numdigits<unsigned>();
+constexpr unsigned INT_MAX_CHARS = type_max_numdigits<int>() + 1;
+constexpr unsigned ULONG_MAX_CHARS = type_max_numdigits<unsigned long>();
+constexpr unsigned LONG_MAX_CHARS = type_max_numdigits<long>() + 1;
+
+} /* anonymous namespace */
+
+namespace dio {
+
+// class io_base
+
+streambuf *io_base::get_buf() noexcept
+{
+    return buf.get();
+}
+
+bool io_base::is_open() noexcept
+{
+    return (fd >= 0);
+}
+
+void io_base::set_fd(const int newfd) noexcept
+{
+    fd = newfd;
+}
+
+int io_base::get_fd() noexcept
+{
+    return fd;
+}
+
+// class ostream
+
+ssize_t ostream::put(const char *msg, size_t count) noexcept
+{
+    if (!good()) {
+        return -1;
+    }
+
+    if (count == 0) {
+        return 0;
+    }
+
+    size_t remaining_bytes = count;
+    while (remaining_bytes > static_cast<unsigned>(buf->get_free())) {
+        struct iovec bufv[3];
+        char *ptr = buf->get_ptr(0);
+
+        bufv[0].iov_base = ptr;
+        bufv[0].iov_len = buf->get_contiguous_length(ptr);
+        if (buf->get_length() != bufv[0].iov_len) {
+            // buffer's data in streambuf is split into two parts in this case because of cpbuffer nature.
+            bufv[1].iov_base = buf->get_buf_base();
+            bufv[1].iov_len = buf->get_length() - bufv[0].iov_len;
+        }
+        else {
+            bufv[1].iov_base = nullptr;
+            bufv[1].iov_len = 0;
+        }
+        bufv[2].iov_base = const_cast<char *>(msg);
+        bufv[2].iov_len = remaining_bytes;
+
+        unsigned buffer_length = bufv[0].iov_len + bufv[1].iov_len;
+
+        ssize_t cur_written_bytes = bp_sys::writev(fd, bufv, 3);
+        if (cur_written_bytes < 0) {
+            io_error = errno;
+            // Some bytes may be written in previous iteration
+            return (count - remaining_bytes > 0) ? count - remaining_bytes : -1;
+        }
+        if (cur_written_bytes >= buffer_length) {
+            buf->consume(buffer_length);
+            cur_written_bytes -= buffer_length;
+            remaining_bytes -= cur_written_bytes;
+            if (remaining_bytes == 0) {
+                return count;
+            }
+            msg += cur_written_bytes;
+        }
+        else {
+            buf->consume(cur_written_bytes);
+        }
+    }
+
+    buf->append(msg, remaining_bytes);
+    return count;
+}
+
+void ostream::throw_exception_on(const int states)
+{
+    if ((states & io_states::buffer_failbit) && buffer_failure()) {
+        throw iostream_internal_err("Cannot use stream's buffer: buffer is nullptr!");
+    }
+    if ((states & io_states::io_failbit) && io_failure()) {
+        throw iostream_system_err(io_failure());
+    }
+}
+
+bool ostream::open_nx(const char *path) noexcept
+{
+    fd = bp_sys::open(path, O_WRONLY);
+
```

Comment:
Why the blank line here? (and not in the equivalent place in the function just below)